Tahi-project/tahi
# APERTA-9951 automate similarity check

JIRA issue: https://jira.plos.org/jira/browse/APERTA-9951

#### How to test this
1. Add a Similarity Check Task to a workflow (a manuscript manager template), I'd recommend starting with the 'No Author Tasks' workflow as it makes actually submitting quick.
2. Run the rake task to enable the similarity check for the given workflow.  For instance, if you added the similarity check to MMT 3, and you want the check to happen at first full submission, you'll run `rake settings:enable_ithenticate_automation[3,at_first_full_submission]`
3. Now upload a new paper for that workflow, complete all the author tasks, and submit.
You should see that the similarity check task in the workflow view is already completed.


#### What this PR does:

This PR does two major things.
* Creates a Setting infrastructure that we can use to assign configuration information to arbitrary stuff in the database (it's less scary than I'm making it sound, hopefully)
* When a paper is submitted, we check the per-worfklow settings for the `TahiStandardTasks::SimilarityCheckTask`, and conditionally run a similarity check for the paper.

#### Notes

- We introduce a generic `Setting` model and table that can hold configuration
information for anything.  Setting belongs to a polymorphic owner.
- A `Configurable` is a thing that has `Setting`s in the db.  `Setting`s are
polymorphic and are also named.  Currently we only have one `Setting` subclass,
but in the future one could imagine a `Configurable` having (for instance) five
Boolean settings, all with different names.
- `Settings` are associated to a `Configurable` by a `SettingTemplate`.
`SettingTemplate` is a table that will be seeded per journal.  `SettingTemplate`
lets us answer the question "what kind of settings does this thing have" in the
case of something like a `TaskTemplate`, where different kinds of settings could
be associated to different kinds of `TaskTemplate`s.  A `TaskTemplate` can give a setting a default value and a set of possible values that the setting could take.  `PossibleSettingValue` is another new model we introduced.
- Although `SettingTemplate`s will be seeded, `Setting`s are intended to be lazily
created as needed since they're sparse (initially they'll only be used by one
type of `TaskTemplate`)


#### Major UI changes

The UI work for this ticket is a part of https://jira.plos.org/jira/browse/APERTA-9495.

---

#### Code Review Tasks:

Author tasks:
(need for data migration obviated by the idempotent `settings:seed_setting_templates` task that runs on deploy and seed and such)

Reviewer tasks:


#### After the Code Review:

Author tasks:
